### PR TITLE
chore: Migration of some components to TypeScript

### PR DESCRIPTION
**Pull request by ryo-rm**
## Description

I have migrated some components to TypeScript. I'm mainly migrating Layer components.

## Checklist

- [x] I have tested this on a device/simulator for each compatible OS
- [x] I updated the documentation with running `yarn generate` in the root folder
- [x] I updated the typings files - if js interface was changed (`index.d.ts`)
- [ ] I added/ updated a sample - if a new feature was implemented (`/example`)

### CODE REVIEW

File: __tests__/components/BackgroundLayer.test.js [hunk 1]
```diff
@@ -42,11 +42,6 @@ describe('BackgroundLayer', () => {
     expect(props.filter).toStrictEqual(testProps.filter);
     expect(props.minZoomLevel).toStrictEqual(testProps.minZoomLevel);
     expect(props.maxZoomLevel).toStrictEqual(testProps.maxZoomLevel);
-    expect(props.reactStyle).toStrictEqual({
```
Comment (mfazekas) [inline __tests__/components/BackgroundLayer.test.js:45]: what happened with this test?

**Review by mfazekas — COMMENTED** [review]
@ryo-rm thanks much, that's amazing! 
Looks good to me for the first glance, will do a couple of tests

File: __tests__/components/BackgroundLayer.test.js (hunk 1 above)
Comment (ryo-rm) [reply __tests__/components/BackgroundLayer.test.js:45]: I had overlooked this code. I will fix it!
https://github.com/rnmapbox/maps/blob/main/javascript/components/AbstractLayer.tsx#L59

Comment (ryo-rm) [reply __tests__/components/BackgroundLayer.test.js:45]: I have added a test.

https://github.com/rnmapbox/maps/pull/2566/commits/7fdd75e3db6d6aa4152fdf92c6771829957474be

**Comment by mfazekas** [comment]
@ryo-rm pls see the lint issues:

```
javascript/components/FillLayer.tsx:2:1 - error TS6133: 'PropTypes' is declared but its value is never read.
[43](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:44)

[44](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:45)
2 import PropTypes from 'prop-types';
[45](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:46)
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[46](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:47)

[47](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:48)
javascript/components/HeatmapLayer.tsx:2:1 - error TS6133: 'PropTypes' is declared but its value is never read.
[48](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:49)

[49](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:50)
2 import PropTypes from 'prop-types';
[50](https://github.com/rnmapbox/maps/actions/runs/4041255447/jobs/6969337160#step:5:51)
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by ryo-rm** [comment]
@mfazekas 
Thanks for the review. Fixed  26fb26c53906265b3a4f48cc51921abd546b6a69

**Comment by mfazekas** [comment]
@ryo-rm pls also see type check related issue:

```
$ npx tsc --noEmit
[6](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:7)
Error: src/examples/SymbolCircleLayer/Earthquakes.tsx(206,15): error TS2769: No overload matches this call.
[7](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:8)
  Overload 1 of 2, '(props: Props | Readonly<Props>): CircleLayer', gave the following error.
[8](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:9)
    Type 'CircleLayerStyle' is not assignable to type '(CircleLayerStyleProps & AllLayerStyleProps) | undefined'.
[9](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:10)
      Type 'CircleLayerStyle' is not assignable to type 'CircleLayerStyleProps & TerrainLayerStyleProps'.
[10](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:11)
        Type 'CircleLayerStyle' is not assignable to type 'CircleLayerStyleProps'.
[11](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:12)
          Types of property 'visibility' are incompatible.
[12](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:13)
            Type 'Expression | Visibility | undefined' is not assignable to type 'Enum<VisibilityEnum, VisibilityEnumValues> | undefined'.
[13](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:14)
              Type 'Expression' is not assignable to type 'Enum<VisibilityEnum, VisibilityEnumValues> | undefined'.
[14](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:15)
  Overload 2 of 2, '(props: Props, context: any): CircleLayer', gave the following error.
[15](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:16)
    Type 'CircleLayerStyle' is not assignable to type '(CircleLayerStyleProps & AllLayerStyleProps) | undefined'.
[16](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:17)
Error: src/examples/SymbolCircleLayer/Earthquakes.tsx(212,15): error TS2769: No overload matches this call.
[17](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:18)
  Overload 1 of 2, '(props: Props | Readonly<Props>): CircleLayer', gave the following error.
[18](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:19)
    Type 'CircleLayerStyle' is not assignable to type '(CircleLayerStyleProps & AllLayerStyleProps) | undefined'.
[19](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:20)
  Overload 2 of 2, '(props: Props, context: any): CircleLayer', gave the following error.
[20](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:21)
    Type 'CircleLayerStyle' is not assignable to type '(CircleLayerStyleProps & AllLayerStyleProps) | undefined'.
[21](https://github.com/rnmapbox/maps/actions/runs/4060111556/jobs/6990940148#step:9:22)
error Command failed with exit code 2.
````

**Review by ryo-rm — COMMENTED** [review]
@mfazekas Thanks! fixed.

**Pull request merged**